pkg/conversion/cli.go

```diff
@@ -0,0 +1,211 @@
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/pkg/errors"
+
+	"github.com/crossplane-contrib/terrajet/pkg/meta"
+	"github.com/crossplane-contrib/terrajet/pkg/terraform/resource"
+	"github.com/crossplane-contrib/terrajet/pkg/tfcli"
+)
+
+const (
+	errCannotConsumeState = "cannot consume state"
+
+	errFmtCannotDoWithTFCli = "cannot %s with tf cli"
+)
+
+// BuildClientForResource returns a tfcli client by setting attributes
+// (i.e. desired spec input) and terraform state (if available) for a given
+// client builder base.
+func BuildClientForResource(builderBase tfcli.Builder, tr resource.Terraformed) (tfcli.Client, error) {
+	var stateRaw []byte
+	if meta.GetState(tr) != "" {
+		stEnc := meta.GetState(tr)
+		st, err := BuildStateV4(stEnc, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "cannot build state")
+		}
+
+		stateRaw, err = st.Serialize()
+		if err != nil {
+			return nil, errors.Wrap(err, "cannot serialize state")
+		}
+	}
+
+	attr, err := tr.GetParameters()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get attributes")
+	}
+
+	return builderBase.WithState(stateRaw).WithResourceBody(attr).BuildClient()
+}
+
+// CLI is an Adapter implementation for Terraform CLI
+type CLI struct {
+	tfcli tfcli.Client
+}
+
+// NewCli returns a CLI object
+func NewCli(client tfcli.Client) *CLI {
+	return &CLI{
+		tfcli: client,
+	}
+}
+
+// Observe is a Terraform CLI implementation for Observe function of Adapter interface.
+func (t *CLI) Observe(ctx context.Context, tr resource.Terraformed) (Observation, error) {
+
+	tfRes, err := t.tfcli.Refresh(ctx, xpmeta.GetExternalName(tr))
+
+	if isOperationInProgress(err, tfcli.OperationApply) {
+		//  A previously started "Apply" operation is in progress or completed
+		//  but one last call needs to be done as completed to be able to kick
+		//  off a new operation. We will return "Exists: true, UpToDate: false"
+		//  in order to trigger an Update call.
+		return Observation{
+			Exists:   true,
+			UpToDate: false,
+		}, nil
+	}
+
+	if isOperationInProgress(err, tfcli.OperationDestroy) {
+		// A previously started "Destroy" operation is in progress or completed
+		// but one last call needs to be done as completed to be able to kick
+		// off a new operation. We will return "Exists: true, UpToDate: true" in
+		// order to trigger a Delete call (given we already have deletion
+		// timestamp set).
+		return Observation{
+			Exists:   true,
+			UpToDate: true,
+		}, nil
+	}
+
+	if err != nil {
+		return Observation{}, errors.Wrapf(err, errFmtCannotDoWithTFCli, "observe")
+	}
+	// No tfcli operation was in progress, our blocking observation completed
+	// successfully, and we have an observation to consume.
+
+	// If resource does not exist, and it was actually deleted, we no longer
+	// need this client (hence underlying workspace) for this resource.
+	if !tfRes.Exists && xpmeta.WasDeleted(tr) {
```

Comment:
If there is an external name when the CR is created but external resource doesn't exist, how do we tell managed reconciler that the resource doesn't exist? It seems like we return `ResourceExists: false` only if the resource is in deletion state but I might be missing something.

Reply:
We return whatever in `tfRes.Exists` at the last return of this `Observe` func. In the case you described, it would be false and we would return `ResourceExists: false`.